cfpb/rental-assistance-finder
# Add all-county filtering and increase testing

Users are finding it confusing when their county does not show up in the county dropdown for their state (currently only counties with programs are shown). This PR updates the county functionality to show all a state's counties, pulled from Census data. It also adds more browser and unit tests and new cypress component testing.

Note: There are a handful of county programs in the data whose names do not match counties in the Census data. This can be handled by adding a county array to those programs and updating the county filtering code to look for county arrays on county as well as city programs.

## Additions

- Add more unit tests for utils.
- Add app status, filtering, and translations integration tests.
- Add Cypress component testing.
- Doc comments for utils
- Counties json file

## Changes

- Show all counties (from Census data) in county dropdown.
- Minor code reorganization.
- Update filter message logic in ResultsNotification.
- Update filter messaging in translation files.
- Update county filtering to look for county array on county programs

## Testing

1. Check out branch and run `yarn start`
2. Run tests
    - `yarn e2e` for integration tests (run `app-status`, `filters`, `smoke`,  and `translation` tests)
    -  `yarn test` for unit tests
    - `yarn components` for component tests
 3. Test app filtering
     - Single geographic result (Connecticut, California + Alpine county)
     - Single tribe result (any tribe)
     - Multiple results, state (California)
     - Multiple results, state + county (California + Alameda county)
     - Multiple results, state + tribe (California + tribe)
4. Verify census county data-based filtering against data
    - `yarn e2e` and run `verify-live-data` test (this grabs the program json file from the server, processes it into state and county maps, and then checks the tool's filter results against the data). Note that there are several errors, where county names don't match values in the county dropdown
    -  `yarn e2e` and run `verify-fixture-data` test. This runs the same test as the live data, but using a new json fixture file that adds county arrays to the counties that errored in the above test. Verify that these tests pass.

## Screenshots

### County dropdown 
Before | After
 --- | ---
<img width="693" alt="Screen Shot 2021-09-29 at 1 05 49 PM" src="https://user-images.githubusercontent.com/778171/135340986-56b8541e-30c3-4a80-9fc5-018e723ea412.png"> | <img width="751" alt="Screen Shot 2021-09-29 at 1 05 28 PM" src="https://user-images.githubusercontent.com/778171/135340954-944e2d73-6e5a-4684-bfce-6510ccc14b4e.png">



### Results messaging

Message type    | Message
 --- | ---
Multiple results, state  | <img width="759" alt="Screen Shot 2021-09-21 at 2 56 56 PM" src="https://user-images.githubusercontent.com/778171/135340298-f5979d5d-4a7d-44e9-b9a3-dadb86d361b0.png">
Multiple results, state&nbsp;+&nbsp;county  | <img width="741" alt="Screen Shot 2021-09-21 at 2 57 10 PM" src="https://user-images.githubusercontent.com/778171/135340335-24135d8a-d36c-47ac-a06b-c79a5fbc508d.png"> 
Single geographic result, state&nbsp;+&nbsp;county  | <img width="751" alt="Screen Shot 2021-09-21 at 2 57 20 PM" src="https://user-images.githubusercontent.com/778171/135339893-226388f5-5c6a-4b84-a423-ca6c1d779bf7.png">
Single geographic result, state  | <img width="751" alt="Screen Shot 2021-09-21 at 2 57 20 PM" src="https://user-images.githubusercontent.com/778171/135340079-22e29752-29fc-4bc3-9952-f12ffa7bf42f.png">
Single geographic result, territory  | <img width="680" alt="Screen Shot 2021-09-21 at 11 51 08 PM" src="https://user-images.githubusercontent.com/778171/135339645-01123c4b-7e76-4852-809a-c718e21496e4.png">
Single tribe result  | <img width="742" alt="Screen Shot 2021-09-21 at 2 58 13 PM" src="https://user-images.githubusercontent.com/778171/135339742-f322ed4b-7202-4fe3-9994-d4a34af04bbc.png">
Multiple results, state + tribe  | <img width="767" alt="Screen Shot 2021-09-21 at 2 58 29 PM" src="https://user-images.githubusercontent.com/778171/135339710-f3722c44-6c14-4d82-a3a3-d8c9922b45c5.png">





## Checklist


## Testing checklist

### Browsers


### Accessibility


### Other
